File: VoiceTranscriber-Windows/Verbalize/Services/PipelineLogger.cs
// Module: Pipeline timing logger — writes CSV + console output for performance analysis
using System.Globalization;

namespace Verbalize.Services;

public static class PipelineLogger
{
    private static readonly string LogDir = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Verbalize");
    private static readonly string LogFile = Path.Combine(LogDir, "pipeline_timing.csv");
    private static readonly object FileLock = new();

    private const string Header = "timestamp,engine,stt_model,transcribe_ms,cleanup_ms,cleanup_method,cleanup_model,audio_duration_s,word_count,total_ms,error";

    static PipelineLogger()
    {
        Directory.CreateDirectory(LogDir);
        if (!File.Exists(LogFile))
            File.WriteAllText(LogFile, Header + Environment.NewLine);
    }

    public static void Log(
        string engine,
        string sttModel,
        int transcribeMs,
        int cleanupMs,
        string cleanupMethod,
        string cleanupModel,
        double audioDuration,
        int wordCount,
        int? totalMs = null,
        string? error = null)
    {
        var total = totalMs ?? (transcribeMs + cleanupMs);
        var escapedError = (error ?? "").Replace(",", ";");
        var timestamp = DateTime.UtcNow.ToString("o");

        // Console/Debug output
        if (!string.IsNullOrEmpty(error))
            System.Diagnostics.Debug.WriteLine($"[Pipeline] [{engine}/{sttModel}] FAILED after {total}ms — {error}");
        else
            System.Diagnostics.Debug.WriteLine($"[Pipeline] [{engine}/{sttModel}] transcribe={transcribeMs}ms cleanup={cleanupMs}ms ({cleanupMethod}/{cleanupModel}) total={total}ms | {audioDuration:F1}s audio → {wordCount} words");

        // CSV append
        var line = $"{timestamp},{engine},{sttModel},{transcribeMs},{cleanupMs},{cleanupMethod},{cleanupModel},{audioDuration.ToString("F1", CultureInfo.InvariantCulture)},{wordCount},{total},{escapedError}";
        lock (FileLock)
        {
            try
            {
                File.AppendAllText(LogFile, line + Environment.NewLine);
            }
            catch
            {
                // Don't crash the app if logging fails
            }
        }
    }

    public static string LogFilePath => LogFile;

    public static List<PipelineEntry> RecentEntries(int count = 20)
    {
        if (!File.Exists(LogFile)) return new();

        try
        {
            var lines = File.ReadAllLines(LogFile)
                .Where(l => !string.IsNullOrWhiteSpace(l) && !l.StartsWith("timestamp"))
                .ToList();
            return lines.TakeLast(count).Reverse().Select(PipelineEntry.Parse).Where(e => e != null).Cast<PipelineEntry>().ToList();
        }
        catch
        {
            return new();
        }
    }
}

public class PipelineEntry
{
    public string Timestamp { get; init; } = "";
    public string Engine { get; init; } = "";
    public string SttModel { get; init; } = "";
    public int TranscribeMs { get; init; }
    public int CleanupMs { get; init; }
    public string CleanupMethod { get; init; } = "";
    public string CleanupModel { get; init; } = "";
    public double AudioDuration { get; init; }
    public int WordCount { get; init; }
    public int TotalMs { get; init; }
    public string? Error { get; init; }

    public string Summary =>
        !string.IsNullOrEmpty(Error)
            ? $"[{Engine}] FAILED — {Error}"
            : $"[{Engine}/{SttModel}] {TranscribeMs}ms + {CleanupMs}ms ({CleanupMethod}/{CleanupModel}) = {TotalMs}ms total | {WordCount} words";

    public static PipelineEntry? Parse(string csvLine)
    {
        var parts = csvLine.Split(',');
        if (parts.Length < 10) return null;

        string timestamp;
        if (DateTimeOffset.TryParse(parts[0], out var dto))
            timestamp = dto.LocalDateTime.ToString("h:mm:ss tt");
        else
            timestamp = parts[0];

        return new PipelineEntry
        {
            Timestamp = timestamp,
            Engine = parts[1],
            SttModel = parts[2],
            TranscribeMs = int.TryParse(parts[3], out var t) ? t : 0,
            CleanupMs = int.TryParse(parts[4], out var c) ? c : 0,
            CleanupMethod = parts[5],
            CleanupModel = parts[6],
            AudioDuration = double.TryParse(parts[7], NumberStyles.Float, CultureInfo.InvariantCulture, out var d) ? d : 0,
            WordCount = int.TryParse(parts[8], out var w) ? w : 0,
            TotalMs = int.TryParse(parts[9], out var m) ? m : 0,
            Error = parts.Length > 10 && !string.IsNullOrWhiteSpace(parts[10]) ? parts[10] : null
        };
    }
}

File: VoiceTranscriber/VoiceTranscriber/Utils/PipelineLogger.swift
import Foundation
import os

/// Logs pipeline timing data for performance analysis.
/// Writes both to os_log (visible in Console.app) and to a CSV file in Application Support.
final class PipelineLogger {
    static let shared = PipelineLogger()

    private let logger = Logger(subsystem: "com.verbalize.app", category: "pipeline")
    private let fileURL: URL
    private let queue = DispatchQueue(label: "com.verbalize.pipelinelogger")

    private init() {
        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        let appDir = appSupport.appendingPathComponent("Verbalize")
        try? FileManager.default.createDirectory(at: appDir, withIntermediateDirectories: true)
        self.fileURL = appDir.appendingPathComponent("pipeline_timing.csv")

        // Write CSV header if file doesn't exist
        if !FileManager.default.fileExists(atPath: fileURL.path) {
            let header = "timestamp,engine,stt_model,transcribe_ms,cleanup_ms,cleanup_method,cleanup_model,audio_duration_s,word_count,total_ms,error\n"
            try? header.write(to: fileURL, atomically: true, encoding: .utf8)
        }
    }

    func log(
        engine: String,
        sttModel: String = "",
        transcribeMs: Int,
        cleanupMs: Int,
        cleanupMethod: String,
        cleanupModel: String = "",
        audioDuration: Double,
        wordCount: Int,
        totalMs: Int? = nil,
        error: String? = nil
    ) {
        let total = totalMs ?? (transcribeMs + cleanupMs)
        let errorStr = error ?? ""

        // Console log (visible in Console.app with filter "pipeline")
        if let error = error {
            logger.error("[\(engine)/\(sttModel)] FAILED after \(total)ms — \(error)")
        } else {
            logger.info("[\(engine)/\(sttModel)] transcribe=\(transcribeMs)ms cleanup=\(cleanupMs)ms (\(cleanupMethod)/\(cleanupModel)) total=\(total)ms | \(String(format: "%.1f", audioDuration))s audio → \(wordCount) words")
        }

        // CSV append
        queue.async { [fileURL] in
            let timestamp = ISO8601DateFormatter().string(from: Date())
            let escapedError = errorStr.replacingOccurrences(of: ",", with: ";")
            let line = "\(timestamp),\(engine),\(sttModel),\(transcribeMs),\(cleanupMs),\(cleanupMethod),\(cleanupModel),\(String(format: "%.1f", audioDuration)),\(wordCount),\(total),\(escapedError)\n"
            if let data = line.data(using: .utf8),
               let handle = try? FileHandle(forWritingTo: fileURL) {
                handle.seekToEndOfFile()
                handle.write(data)
                handle.closeFile()
            }
        }
    }

    /// Returns the path to the CSV log file for display in settings
    var logFilePath: String {
        fileURL.path
    }

    /// Returns recent log entries as formatted strings for in-app display
    func recentEntries(count: Int = 20) -> [PipelineEntry] {
        guard let content = try? String(contentsOf: fileURL, encoding: .utf8) else { return [] }
        let lines = content.components(separatedBy: "\n").filter { !$0.isEmpty && !$0.hasPrefix("timestamp") }
        return lines.suffix(count).reversed().compactMap { PipelineEntry(csvLine: $0) }
    }
}

struct PipelineEntry: Identifiable {
    let id = UUID()
    let timestamp: String
    let engine: String
    let sttModel: String
    let transcribeMs: Int
    let cleanupMs: Int
    let cleanupMethod: String
    let cleanupModel: String
    let audioDuration: Double
    let wordCount: Int
    let totalMs: Int
    let error: String?

    init?(csvLine: String) {
        let parts = csvLine.components(separatedBy: ",")
        guard parts.count >= 10 else { return nil }
        if let date = ISO8601DateFormatter().date(from: parts[0]) {
            let fmt = DateFormatter()
            fmt.dateFormat = "h:mm:ss a"
            self.timestamp = fmt.string(from: date)
        } else {
            self.timestamp = parts[0]
        }
        self.engine = parts[1]
        self.sttModel = parts[2]
        self.transcribeMs = Int(parts[3]) ?? 0
        self.cleanupMs = Int(parts[4]) ?? 0
        self.cleanupMethod = parts[5]
        self.cleanupModel = parts[6]
        self.audioDuration = Double(parts[7]) ?? 0
        self.wordCount = Int(parts[8]) ?? 0
        self.totalMs = Int(parts[9]) ?? 0
        self.error = parts.count > 10 && !parts[10].isEmpty ? parts[10] : nil
    }

    var summary: String {
        if let error = error {
            return "[\(engine)] FAILED — \(error)"
        }
        return "[\(engine)/\(sttModel)] \(transcribeMs)ms + \(cleanupMs)ms (\(cleanupMethod)/\(cleanupModel)) = \(totalMs)ms total | \(wordCount) words"
    }
}
